File: services/23-laravel/app/Http/Middleware/RateLimit.php
<?php

namespace App\Http\Middleware;

use Closure;
use Illuminate\Cache\RateLimiter;
use Illuminate\Http\Request;
use Illuminate\Http\Response;
use Symfony\Component\HttpFoundation\Response as SymfonyResponse;

/**
 * Rate limiter middleware — 100 requests per minute per IP.
 * Uses Laravel's built-in RateLimiter backed by the cache driver.
 * Health and docs endpoints are exempt so k8s probes are never blocked.
 */
class RateLimit
{
    public function __construct(private RateLimiter $limiter) {}

    public function handle(Request $request, Closure $next): SymfonyResponse
    {
        // Health and docs endpoints are exempt.
        if (
            str_starts_with($request->path(), 'health') ||
            str_starts_with($request->path(), 'api/documentation') ||
            str_starts_with($request->path(), 'docs')
        ) {
            return $next($request);
        }

        $key = 'rate_limit:' . $request->ip();

        if ($this->limiter->tooManyAttempts($key, 100)) {
            return response()->json(
                ['error' => 'Too many requests — try again in 60 seconds'],
                429
            );
        }

        $this->limiter->hit($key, 60);

        return $next($request);
    }
}

File: services/api-phoenix/lib/app_web/plugs/rate_limit.ex
defmodule AppWeb.Plugs.RateLimit do
  @moduledoc """
  Rate limiter plug — 100 requests per minute per remote IP using Hammer.

  Health and docs endpoints are exempt so k8s probes are never blocked.
  Rejected requests receive HTTP 429 with a JSON error body.
  """

  import Plug.Conn

  # 100 requests per 60 seconds per IP.
  @limit 100
  @scale_ms 60_000

  @exempt_prefixes ["/health", "/api/openapi"]

  def init(opts), do: opts

  def call(%Plug.Conn{request_path: path} = conn, _opts) do
    if Enum.any?(@exempt_prefixes, &String.starts_with?(path, &1)) do
      conn
    else
      ip = conn |> get_peer_data() |> Map.get(:address) |> :inet.ntoa() |> to_string()
      case Hammer.check_rate("rate_limit:#{ip}", @scale_ms, @limit) do
        {:allow, _count} ->
          conn
        {:deny, _limit} ->
          conn
          |> put_resp_content_type("application/json")
          |> send_resp(429, ~s({"error":"Too many requests — try again in 60 seconds"}))
          |> halt()
      end
    end
  end
end
